File: ruby/de_optimizer.rb
require 'pp'

class DE_Optimizer

  class Domain
    attr_reader :min, :max
    def initialize(h)
      @min, @max = h[:min], h[:max]
      raise "invalid range : [#{@min}, #{@max}]" if @min > @max
    end

    def scale(r)    # give [0,1] value and return a value scaled in [min,max]
      r * (@max - @min) + @min
    end
  end

  attr_reader :best_point, :best_f, :t, :population
  attr_accessor :map_func

  def initialize( map_func, domains, n: nil, f: 0.8, cr: 0.9, rand_seed: nil )
    @n, @f, @cr = (n || domains.size*10), f, cr
    @rng = Random.new( rand_seed || Random.new_seed )

    @domains = domains.map {|h| Domain.new(h) }
    @map_func = map_func
    @t = 0
    @best_point = nil
    @best_f = Float::INFINITY

    generate_initial_points
  end

  def generate_initial_points
    @population = Array.new(@n) {|i| @domains.map {|d| d.scale( @rng.rand ) } }
    @current_fs = @map_func.call( @population )
  end

  def average_f
    @current_fs.inject(:+) / @current_fs.size
  end

  def proceed
    new_positions = []
    @n.times do |i|
      new_pos = generate_candidate(i)
      new_positions << new_pos
    end

    new_fs = @map_func.call( new_positions )

    # selection
    @n.times do |i|
      if new_fs[i] < @current_fs[i]
        @population[i] = new_positions[i]
        @current_fs[i] = new_fs[i]
        if new_fs[i] < @best_f
          @best_point = new_positions[i]
          @best_f = new_fs[i]
        end
      end
    end

    @t += 1
  end

  private

  # generate a candidate for @population[i]
  # based on DE/rand/1/binom algorithm
  def generate_candidate(i)
    # randomly pick a,b,c
    begin
      a = @rng.rand( @n )
    end while ( a == i )
    begin
      b = @rng.rand( @n )
    end while ( b == i || b == a )
    begin
      c = @rng.rand( @n )
    end while ( c == i || c == a || c == b )

    # compute the new position
    new_pos = @population[i].dup

    # pick a random index r
    dim = @domains.size
    r = @rng.rand( dim )

    dim.times do |d|
      if( d == r || @rng.rand < @cr )
        new_pos[d] = @population[a][d] + @f * (@population[b][d] - @population[c][d])
      end
    end
    new_pos
  end
end

if $0 == __FILE__
  domains = [
    {min: -10.0, max: 10.0},
    {min: -10.0, max: 10.0},
    {min: -10.0, max: 10.0}
  ]
  f = lambda {|x| (x[0]-1.0)**2+(x[1]-2.0)**2+(x[2]-3.0)**2 }
  map_agents = lambda {|points| points.map(&f) }

  opt = DE_Optimizer.new(map_agents, domains, n: 30, f: 0.8, cr: 0.9, rand_seed: 1234)

  20.times do |t|
    opt.proceed
    puts "#{opt.t} #{opt.best_point} #{opt.best_f} #{opt.average_f}"
  end
end

File: python/de_optimizer.py
import random

class Domain():
    def __init__(self, minimum, maximum):
        self.min = minimum
        self.max = maximum
        assert self.min < self.max

    def scale(self, r):
        """
        give [0,1] value and returns the scaled value
        """
        return r * (self.max - self.min) + self.min

class DE_Optimizer():

    def __init__( self, map_func, domains, n=None, f=0.8, cr=0.9, rand_seed=None ):
        self.n = (n or len(domains)*10)
        self.f = f
        self.cr = cr
        self.random = random.Random()
        if rand_seed:
            self.random.seed( rand_seed )
        self.domains = [ Domain(d['min'],d['max']) for d in domains ]
        self.map_func = map_func
        self.t = 0
        self.best_point = None
        self.best_f = float('inf')

        self.generate_initial_points()

    def generate_initial_points(self):
        self.population = []
        for i in range(self.n):
            point = [ d.scale( self.random.random() ) for d in self.domains ]
            self.population.append( point )
        self.current_fs = self.map_func( self.population )

    def average_f(self):
        return sum( self.current_fs ) / len( self.current_fs )

    def proceed(self):
        new_positions = []
        for i in range(self.n):
            new_pos = self._generate_candidate(i)
            new_positions.append( new_pos )

        new_fs = self.map_func( new_positions )

        # selection
        for i in range(self.n):
            if new_fs[i] < self.current_fs[i]:
                self.population[i] = new_positions[i]
                self.current_fs[i] = new_fs[i]
                if new_fs[i] < self.best_f:
                    self.best_point = new_positions[i]
                    self.best_f = new_fs[i]

        self.t += 1

    def _generate_candidate(self, i):
        """
        generate a candidate for population[i]
        based on DE/rand/1/binom algorithm
        """

        a = i
        while a == i:
            a = self.random.randrange(self.n)
        b = i
        while b == i or b == a:
            b = self.random.randrange(self.n)
        c = i
        while c == i or c == a or c == b:
            c = self.random.randrange(self.n)

        new_pos = self.population[i].copy()

        dim = len(self.domains)
        r = self.random.randrange( dim )

        for d in range(dim):
            if d == r or self.random.random() < self.cr:
                new_pos[d] = self.population[a][d] + self.f * (self.population[b][d] - self.population[c][d])
                if new_pos[d] > self.domains[d].max:
                    new_pos[d] = self.domains[d].max;
                if new_pos[d] < self.domains[d].min:
                    new_pos[d] = self.domains[d].min;
        return new_pos

if __name__ == "__main__":
    def main():
        domains = [
                {'min': -10.0, 'max': 10.0},
                {'min': -10.0, 'max': 10.0},
                {'min': -10.0, 'max': 10.0}
                ]
        def f(x):
            return (x[0]-1.0)**2+(x[1]-2.0)**2+(x[2]-3.0)**2
        def map_agents(points):
            return [ f(p) for p in points ]

        opt = DE_Optimizer(map_agents, domains, n=30, f=0.8, cr=0.9, rand_seed=1234)

        for t in range(20):
            opt.proceed()
            print("t=%d  %s, %f, %f" % (t, repr(opt.best_point), opt.best_f, opt.average_f() ) )

    main()
